Do not prevent Duplicate Notifications (edited)

jjeffryes opened the pull request:
This PR has been changed, now it removes all code that was hiding duplicate notifications.

jjeffryes committed:
Removes all duplicate follow notifications.

rmisio commented:
Can we not just filter out the dupes in the collection in parse? It seems like it could be problematic to have the UI out of sync with the collection.

jjeffryes commented:
I'm going to remove the handling of duplicate follow messages in the notifications (the websocket code will remain, but will be updated).

After doing some more work on it, trying to fix the get_notifications API data in the client causes numerous issues. 

The unread count becomes inaccurate, and since the unread count can include follow notifications that haven't been fetched yet, and the client can't fix that (ie: there are 11 unread duplicate follow messages, but the client only knows about the 6 duplicates it has fetched). 

It also results in some very weird behavior where if you're being spammed, and the client fetched 6 duplicate notifications, only 1 is shown. We'd have to loop the fetch to keep fetching more notifications until we had at least 6 non-duplicates, which seems like a very bad idea.

jjeffryes committed:
Remove handling of duplicate follow notifications, and change the websocket handling to only block duplicate follow messages once an hour, instead of all duplicate messages.

Also renames notificationsMd to notificationMd, since it is the model for one notification.

jjeffryes committed:
Fix bug due to removing too much code.

rmisio commented on js/views/pageNavVw.js:
```diff
@@ -238,6 +238,17 @@ module.exports = baseVw.extend({
       notif.type = notif.type || notif.message_type;
       notifStamp = Date.now();
       notif.guid = notif.guid || notif.sender;
+
```
Even with this approach, the counts will still be off, because when the user lazy loads a new page of notifications, the returned count includes the dupes. You could keep track of yet another offset to account for them, but if I'm remembering correctly we are already keeping track of 3 offsets and it took a lot of tinkering to account for all the edge cases. If you do want to add another offset, please do so at your own risk (I don't recommend it as we've already had weird count related bugs)!

FWIW, IMHO this should really be the serving preventing too frequent dupes on the follow POST so the dupes don't even make it into the DB and therefore never affect the count.

jjeffryes committed:
Don't prevent any duplications.

rmisio commented on js/views/pageNavVw.js:
```diff
@@ -225,18 +224,16 @@ module.exports = baseVw.extend({
       username = notif.sender ? notif.sender : username;
       avatarHash = notif.image_hash || notif.avatar_hash;
       avatar = avatarHash ? app.serverConfigs.getActive().getServerBaseUrl + '/get_image?hash=' +
-        avatarHash + '&guid=' + notif.guid : 'imgs/defaultUser.png';
```
can we keep the leading spaces, since the second line is an extension of the first one?

jjeffryes replied:
Oops. Didn't mean to change that.

jjeffryes committed:
Fix spacing.
